Mods/src/PuppiMod.cc

```diff
@@ -141,235 +160,222 @@ void PuppiMod::SlaveBegin()
-          numFCHPUis0[i1]++;                                        // Count particles to ignore when taking the median
-        if (alphaC[i0] <= 0)
-          numCCHPUis0[i1]++;
-        alphaFCHPU[i1][numCHPU[i1]] = alphaF[i0];                   // Only intializing particles up to the number of charged PU
-        alphaCCHPU[i1][numCHPU[i1]] = alphaC[i0];
-        numCHPU[i1]++;                                              // Count the total number of charged PU particles
+
+        if (alphaF[iCand] <= 0)
+          numFCHPUis0[iEtaBin]++;                                        // Count particles to ignore when taking the median
+
+        if (alphaC[iCand] <= 0)
+          numCCHPUis0[iEtaBin]++;
+
+        alphaFCHPU[iEtaBin].push_back(alphaF[iCand]);                   // Only intializing particles up to the number of charged PU
+        alphaCCHPU[iEtaBin].push_back(alphaC[iCand]);
       }
     }
-    else if (iParticleType == 1 && fApplyLowPUCorr) {               // if charged PV, and trying to correct
+    else if (iParticleType == kChargedPrimary && fApplyLowPUCorr) {               // if charged PV, and trying to correct
       Double_t checkEta = fabs(iCandidate->Eta());
-      for (Int_t i1 = 0; i1 < fNumEtaBins; i1++) {
-        if (checkEta > fEtaMaxExtraps[i1])
+      for (UInt_t iEtaBin = 0; iEtaBin < fNumEtaBins; iEtaBin++) {
+        if (checkEta > fEtaMaxExtraps[iEtaBin])
           continue;                                                 // if outside the binning that we are interested in, don't use CHPV
-        alphaFCHPV[i1][numCHPV[i1]] = alphaF[i0];                   // Only intializing particles up to the number of charged PV
-        alphaCCHPV[i1][numCHPV[i1]] = alphaC[i0];
-        numCHPV[i1]++;                                              // Count the total number of charged PV particles
+
+        alphaFCHPV[iEtaBin].push_back(alphaF[iCand]);                   // Only intializing particles up to the number of charged PV
+        alphaCCHPV[iEtaBin].push_back(alphaC[iCand]);
       }
     }
   }
 
-  // Set the indices for sorting for the charged PU alphas
-  for (Int_t i0 = 0; i0 < fNumEtaBins; i0++) {
-    for (Int_t i1 = 0; i1 < numCandidates; i1++) {
-      IndicesFCHPU[i0][i1] = i1;
-      IndicesCCHPU[i0][i1] = i1;
-    }
+  std::vector<UInt_t> numCHPU(fNumEtaBins);
+  std::vector<UInt_t> numCHPV(fNumEtaBins);
+  for (unsigned iEtaBin = 0; iEtaBin != fNumEtaBins; ++iEtaBin) {
+    numCHPU[iEtaBin] = alphaFCHPU[iEtaBin].size();
+    numCHPV[iEtaBin] = alphaFCHPV[iEtaBin].size();
   }
 
-  // Sort everything to find the median and drop particles faster
-  // This gives back the Indices arrays in ordered for m
-  TMath::Sort(numCandidates,alphaF,IndicesF,0);
-  TMath::Sort(numCandidates,alphaC,IndicesC,0);
-  for (Int_t i0 = 0; i0 < fNumEtaBins; i0++) {
-    TMath::Sort(numCHPU[i0],alphaFCHPU[i0],IndicesFCHPU[i0],0);
-    TMath::Sort(numCHPU[i0],alphaCCHPU[i0],IndicesCCHPU[i0],0);
```

Comment:
Also, TMath::Sort using C++ arrays is way more roundabout than simple std::sort using std::vectors.